### PR TITLE
Add multiselect support for array fields

### PR DESCRIPTION
**Pull request by cedricmessiant**
See #41

### CODE REVIEW

File: test/index_test.js [hunk 1]
```diff
@@ -443,67 +443,135 @@ describe("Form", () => {
...
+        },
+        uniqueItems: true,
+        value: ["foo", "fuzz"]
+      };
+
+      it("should render a select widget", () => {
+        const {node} = createComponent({schema});
+
+        expect(node.querySelectorAll("select"))
+        .to.have.length.of(1);
```
Comment (n1k0) [inline test/index_test.js]: Nit: indentation

File: test/index_test.js [hunk 2]
```diff
@@ -443,67 +443,135 @@ describe("Form", () => {
...
+
+        expect(node.querySelector(".field select").getAttribute("multiple"))
+          .not.to.be.null;
+      });
+
+      it("should render options", () => {
+        const {node} = createComponent({schema});
+
+        expect(node.querySelectorAll("select option"))
+        .to.have.length.of(3);
```
Comment (n1k0) [inline test/index_test.js]: Nit: indentation

File: src/components/fields/ArrayField.js [hunk 3]
```diff
@@ -60,11 +61,31 @@ class ArrayField extends Component {
     }, {validate: false});
   }
 
+  onSelectChange(value) {
+    this.asyncSetState({
+      items: value
+    }, {validate: false});
```
Comment (n1k0) [inline src/components/fields/ArrayField.js]: Nit: could be written on a single line:

```js
onSelectChange(items) {
  this.asyncSetState({items}, {validate: false});
}
```

File: src/components/fields/ArrayField.js [hunk 4]
```diff
@@ -60,11 +61,31 @@ class ArrayField extends Component {
...
+
   render() {
     const {schema, uiSchema, name} = this.props;
     const title = schema.title || name;
     const {items} = this.state;
     const SchemaField = this.props.registry.SchemaField;
+    if (isMultiSelect(schema)) {
+      return (
+        <SelectWidget
+          multiple
```
Comment (n1k0) [inline src/components/fields/ArrayField.js]: TIL this was possible, very cool!

File: src/components/fields/SchemaField.js [hunk 5]
```diff
@@ -67,11 +70,20 @@ Wrapper.defaultProps = {
...
+  if (schema.type === "object") {
+    displayLabel = false;
+  }
+
   return (
     <Wrapper
       label={schema.title || name}
       required={required}
       type={schema.type}
+      displayLabel={ displayLabel }
```
Comment (n1k0) [inline src/components/fields/SchemaField.js]: Nit: unwanted spaces within

File: src/components/widgets/SelectWidget.js [hunk 6]
```diff
@@ -23,15 +23,30 @@ function SelectWidget({
   value,
   defaultValue,
   required,
+  multiple,
   onChange
 }) {
   return (
     <select
+      multiple={multiple}
```
Comment (n1k0) [inline src/components/widgets/SelectWidget.js]: So that could be just `multiple`, right?

File: src/components/widgets/SelectWidget.js [hunk 7]
```diff
@@ -23,15 +23,30 @@ function SelectWidget({
...
       onChange={(event) => {
-        onChange(processValue(schema.type, event.target.value));
+        let newValue;
+        if (multiple) {
+          newValue = [];
+          const options = event.target.options;
+          for (let i = 0; i < options.length; i++) {
+            if (options[i].selected) {
+              newValue.push(options[i].value);
+            }
```
Comment (n1k0) [inline src/components/widgets/SelectWidget.js]: Nit: How about using `map` and `filter` here?

File: test/index_test.js [hunk 8]
```diff
@@ -443,67 +443,135 @@ describe("Form", () => {
...
+      };
 
-    it("should render a fieldset", () => {
-      const {node} = createComponent({schema});
+      it("should render a fieldset", () => {
+        const {node} = createComponent({schema});
 
-      expect(node.querySelectorAll("fieldset"))
+        expect(node.querySelectorAll("fieldset"))
         .to.have.length.of(1);
```
Comment (n1k0) [inline test/index_test.js]: Nit: indentation

File: test/index_test.js [hunk 9]
```diff
@@ -443,67 +443,135 @@ describe("Form", () => {
...
+      });
 
-    it("should render a fieldset legend", () => {
-      const {node} = createComponent({schema});
+      it("should render a fieldset legend", () => {
+        const {node} = createComponent({schema});
 
-      expect(node.querySelector("fieldset > legend").textContent)
+        expect(node.querySelector("fieldset > legend").textContent)
         .eql("my list");
```
Comment (n1k0) [inline test/index_test.js]: Nit: indentation

File: test/index_test.js [hunk 10]
```diff
@@ -443,67 +443,135 @@ describe("Form", () => {
...
+      });
 
-    it("should contain no field in the list by default", () => {
-      const {node} = createComponent({schema});
+      it("should contain no field in the list by default", () => {
+        const {node} = createComponent({schema});
 
-      expect(node.querySelectorAll(".field-string"))
+        expect(node.querySelectorAll(".field-string"))
         .to.have.length.of(0);
```
Comment (n1k0) [inline test/index_test.js]: Nit: indentation

File: test/index_test.js [hunk 11]
```diff
@@ -443,67 +443,135 @@ describe("Form", () => {
...
+      });
 
-    it("should have an add button", () => {
-      const {node} = createComponent({schema});
+      it("should have an add button", () => {
+        const {node} = createComponent({schema});
 
-      expect(node.querySelector(".array-item-add button"))
+        expect(node.querySelector(".array-item-add button"))
         .not.eql(null);
```
Comment (n1k0) [inline test/index_test.js]: Ditto

File: test/index_test.js [hunk 12]
```diff
@@ -443,67 +443,135 @@ describe("Form", () => {
...
-      const {node} = createComponent({schema});
+      it("should add a new field when clicking the add button", () => {
+        const {node} = createComponent({schema});
 
-      Simulate.click(node.querySelector(".array-item-add button"));
+        Simulate.click(node.querySelector(".array-item-add button"));
 
-      expect(node.querySelectorAll(".field-string"))
+        expect(node.querySelectorAll(".field-string"))
         .to.have.length.of(1);
```
Comment (n1k0) [inline test/index_test.js]: Ditto

File: test/index_test.js [hunk 13]
```diff
@@ -443,67 +443,135 @@ describe("Form", () => {
   });
 
   describe("ArrayField", () => {
-    const schema = {
-      type: "array",
-      title: "my list",
-      items: {type: "string"}
-    };
+    describe("List of inputs", () => {
```
Comment (n1k0) [inline test/index_test.js]: Nit: `"Multiple choices list"`?

File: test/index_test.js [hunk 14]
```diff
@@ -443,67 +443,135 @@ describe("Form", () => {
...
+      it("should handle a change event", () => {
+        const {comp, node} = createComponent({schema});
+
+        Simulate.change(node.querySelector(".field select"), {
+          target: {options: [
+            {selected: true, value: "foo"},
+            {selected: true, value: "bar"},
+            {selected: false, value: "fuzz"},
+          ]
+          }
```
Comment (n1k0) [inline test/index_test.js]: Nit: indentation

**Comment by n1k0** [comment]
Man, this is awesome! Could you please write some docs, and possible update the demo app with some usage of this patch?

Awesome!

File: src/components/widgets/SelectWidget.js (hunk 7 above)
Comment (cedricmessiant) [inline src/components/widgets/SelectWidget.js]: Unfortunately not, options is not `array` !

File: src/components/widgets/SelectWidget.js (hunk 6 above)
Comment (cedricmessiant) [inline src/components/widgets/SelectWidget.js]: No, when you use `multiple`, it means `multiple={ true }` (it is not the same as in objects)

**Comment by almet** [comment]
That's great, big props for the pull request @cedricmessiant !

File: src/components/widgets/SelectWidget.js (hunk 7 above)
Comment (n1k0) [inline src/components/widgets/SelectWidget.js]: Ỳou sure? `for (let i = 0; i < options.length; i++)` sounds like it is one.

Comment (n1k0) [inline src/components/widgets/SelectWidget.js]: Note: If it's matter of it not fully implementing the Array interface, you can simply `[].call(options).map(...)`.

Comment (cedricmessiant) [inline src/components/widgets/SelectWidget.js]: Okay, I'll try this ! (I first tried to use filter and map)

**Comment by n1k0** [comment]
This is fantastic, you rock.

**Comment by n1k0** [comment]
Version 0.12.0 published to npm, demo pushed to gh-pages. Thanks again.

**Comment by cedricmessiant** [comment]
You're welcome ! And thanks for the version on npm

**Pull request merged**